### PR TITLE
mqtt: module to support mqtt pubsub messaging

### PR DESCRIPTION
**Pull request by towe75**
#### Pre-Submission Checklist

- [x] Commit message has the format required by CONTRIBUTING guide
- [x] Commits are split per component (core, individual modules, libs, utils, ...)
- [x] Each component has a single commit (if not, squash them into one commit)
- [ ] No commits to README files for modules (changes must be done to docbook files
in `doc/` subfolder, the README file is autogenerated)

#### Type Of Change
- [ ] Small bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)

#### Checklist:

- [ ] PR should be backported to stable branches
- [x] Tested changes locally
- [ ] Related to issue #XXXX (replace XXXX with an open issue number)

#### Description

New Module to participate in a MQTT pub/sub messaging system. 

Initial Features:

- automatic reconnect with connected/disconnected event routes
- support mqtt last will feature
- subscribe and unsubscribe topics via script code
- receive messages via event route
- send messages from script code
- full KEMI support

### CODE REVIEW

**Comment by oej** [comment]
I would love to see TLS support clearly implemented and documented. After all, it's 2020 :-)

File: src/modules/mqtt/README [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+event_route[mqtt:message] {
+        xlog("mqtt message received: $mqtt(topic) -> $mqtt(msg) !\n");
+}
+...
+
+6. Exported pseudo-variables
+
+     * $mqtt(topic) - Received topic (only in mqtt:message)
+     * $mqtt(msg) - Received message (only in mqtt:message)
+
```
Comment (oej) [inline src/modules/mqtt/README]: How do I reach the QoS value for the message?

**Comment by oej** [comment]
Thank you for this pull request. Generally, I miss these improvements to your code:
- TLS
- Failover (multiple MQTT servers)

Maybe one could want to set event_route per subscription too.

File: src/modules/mqtt/README (hunk 1 above)
Comment (towe75) [reply src/modules/mqtt/README]: @oej  it's not exposed ATM but it's simple to add it. I will enhance the module accordingly.

**Comment by towe75** [comment]
@oej thank you for the **really** quick review (only a couple of minutes after submission). 

I will check the effort for TLS support. It should not be very high given the fact that libmosquitto exposes a highlevel API for it. 

Regarding event_route per subscription: it's a nice idea and saves the script author a bit of boilerplate code. On the other hand: it would have to support pattern matching for wildcard topics to be really useful. I will not add it for now because of expected complexity.

**Comment by oej** [comment]
Regarding event_route per subscription: I was toying with the idea to add a new parameter to the subscribe function. I think you consider something more automatic, but that would be more complex. Why not just add a third parameter?

**Comment by towe75** [comment]
@oej: it would IMHO only be simple for "static" subscriptions like "kamailio/foo" -> on_mqtt_foo(). 

But what about wildcard subscriptions like "kamailio/+/unregister" or "kamailio/#" ? I can, of course, forbit these mappings by some validation rule but a cleaner solution would be to pattern-match each incoming message against the wildcard subscription to find the correct route. 

It's IMHO a somewhat sophisticated feature for a later version, if the need arises. For now the mapping for incoming messages can be done in script code. We can discuss about it in more detail on the mailing list or some other channel, if you want.

**Comment by oej** [comment]
Ok, let's add new features later and polish this module. It's a very good start compared with my core-dumping first attempt :-)

Will try to test it soon.

File: src/modules/mqtt/README (hunk 1 above)
Comment (towe75) [reply src/modules/mqtt/README]: @oej, see commit ea69941

**Comment by towe75** [comment]
I added TLS support in commit 81852e5. The documentation is enhanced as well. 

It's now possible to have a plain connection as well as TLS with configurable method, ciphers, client certificate and - if desired - disabled server cert checks. 

Parameters are more or less directly forwarded to libmosquitto and subsequently into openssl. So the default optionset depends on the installed openssl version and the available libmosquitto version.

**Comment by miconda** [comment]
Going to merge it, is not a lot of code and further reviews and adjustments can be done directly in the repo afterwards. I will grant @towe75 git commit access so he can maintain the module. Thanks for the contribution!

**Pull request merged**